src/TreeView/usePageUpDown.ts

```diff
@@ -0,0 +1,147 @@
+                return false
+              }
+
+              const position = document.activeElement?.compareDocumentPosition(item)
+              if (position === undefined) {
+                return false
+              }
+
+              return !!(position & Node.DOCUMENT_POSITION_PRECEDING)
+            })
+            .reverse()
+
+          for (const item of items) {
+            if (item.firstElementChild === null) {
+              continue
+            }
+
+            // Check the first element element instead of the item in case the
+            // tree contains child items
+            if (isInViewport(item.firstElementChild)) {
+              item.focus()
+              break
+            }
+          }
+        })
+      }
+
+      // When pressing PageDown, look at treeitem's below the current active element
+      // and find the first one that is not in the viewport. If one exists, focus it
+      if (event.key === 'PageDown') {
+        onScrollEnd(() => {
+          const items: HTMLElement[] = Array.from(tree.querySelectorAll('[role="treeitem"]')).filter(
+            (item): item is HTMLElement => {
+              if (item.nodeType !== Node.ELEMENT_NODE) {
+                return false
+              }
+
+              const position = document.activeElement?.compareDocumentPosition(item)
+              if (position === undefined) {
+                return false
+              }
+
+              return !!(position & Node.DOCUMENT_POSITION_FOLLOWING)
+            }
+          )
+
+          for (const item of items) {
+            if (item.firstElementChild === null) {
+              continue
+            }
+
+            // Check the first element element instead of the item in case the
+            // tree contains child items
+            if (isInViewport(item.firstElementChild)) {
+              item.focus()
+              break
+            }
+          }
+        })
+      }
+    }
+
+    tree.addEventListener('keydown', listener)
+
+    return () => {
+      tree.removeEventListener('keydown', listener)
+    }
+  }, [ref, onScrollEnd])
+}
+
+/**
+ * Provides a function, `onScrollEnd`, that will call the given callback when it
+ * detects that the application is no longer scrolling.
+ */
+function useScrollEnd() {
+  // Keep track of the current timeout and scroll listener in case the owner
+  // component unmounts before the timer has a chance to complete
+  const timeoutId = React.useRef<number | null>(null)
+  const activeListener = React.useRef<(() => void) | null>(null)
+  const onScrollEnd = React.useCallback(callback => {
+    function listener() {
+      if (timeoutId.current !== null) {
+        clearTimeout(timeoutId.current)
+      }
+      timeoutId.current = window.setTimeout(() => {
+        callback()
+        window.removeEventListener('scroll', listener)
+        activeListener.current = null
+      }, 60)
+    }
+
+    window.addEventListener('scroll', listener)
```

Comment:
Can we attach this listener to the nearest scroll container instead of `window`?

It looks like the nested scroll container story in storybook doesn't work:
https://user-images.githubusercontent.com/4608155/200848391-bc13c325-9366-45de-8757-4e5c5619c24e.mp4

Reply:
Theres a `getScrollContainer` function in PageLayout that we might be able to reuse here: https://github.com/primer/react/blob/dc1cf6da2864dd7dfa4ed0cb2c5362f9ff9dd60f/src/PageLayout/useStickyPaneHeight.ts#L109-L115

Reply:
Done! Thanks for the suggesting, pulled it out into a utils file.